### PR TITLE
Replaced 'Advanced Options' toggle with system to select additional inputs per user's situation.

### PR DESCRIPTION
**Pull request by brian-courts**
The major changes here attempt to address two possible issues:
1) Some users might not know when to select 'Advanced Options'. For example, a user might be eligible for a pension from employment not covered by Social Security taxes, but not consider that relevant to their Social Security.
2) Some users might be overwhelmed by the multitude of inputs required when 'Advanced Options' is selected.

The revised system lets the user select only the additional inputs that apply to their situation.

There are also a few minor revisions unrelated to the above issues, including some clarification of the source and value of the default discount rate. 

WARNING: While revising home.component.html, in an attempt to better organize my revisions, I made the mistake of formatting the document. That resulted in some changes in leading and trailing whitespace. I don't find a way to tell GitHub to ignore those insignificant changes, as I can with VSCode. There are many significant changes in that file, but GitHub shows some bulk changes which don't really exist.

### CODE REVIEW

**Comment by MikePiper** [comment]
Firstly: this is really, really cool. A marked improvement in usability. Thank you once again for taking the time to create improvements.

I'm testing it out and going over the code, and my only concerns are about discount rate. I'm not sure I understand why we want the message about the source for discount rate in the user interface, other than simply stating where the default comes from. This feels confusing to me.

Relatedly, why do we want a default discount rate in the URL? Why would a user want to specify not only the discount rate they want to use, but also a default discount rate (which then is just overridden)?

**Comment by brian-courts** [comment]
I'm very glad that you find the changes useful improvements.

Regarding the discount rate, I simply expanded some on the input explanation you provide in the current online version, which is as follows:
"A "discount rate" is necessary to reflect the fact that a dollar today is worth more than a dollar in the future, because a dollar today can be invested. The default here is the yield on 20-year TIPS. See this article for a discussion of why that is chosen as the default." 

I modified it to 
(1) fit conditions these days, when the TIPS rate is negative and a dollar today is not worth more than a dollar in the future.
(2) specify the source, when it is not the TIPS rate.

If the now-longer explanation is confusing, it might be appropriate to add a Discount Rate heading and explanation in the About page, with a link in the user interface to that fragment.

Regarding the default discount rate in the URL: A user might find it helpful to be aware that the person who generated the URL chose not to use the default (probably TIPS) rate. 

Come to think of it, if a user gets the rate from a URL, we could to show the URL rates AND the TIPS rate (if it is available). The URL selected rate would be the input value, but the TIPS rate would be available for reference. That would be a minor revision to the code.

Brian Courts
     On Monday, July 13, 2020, 01:34:21 PM CDT, MikePiper <notifications@github.com> wrote:  
 
 


Firstly: this is really, really cool. A marked improvement in usability. Thank you once again for taking the time to create improvements.

I'm testing it out and going over the code, and my only concerns are about discount rate. I'm not sure I understand why we want the message about the source for discount rate in the user interface, other than simply stating where the default comes from. This feels confusing to me.

Relatedly, why do we want a default discount rate in the URL? Why would a user want to specify not only the discount rate they want to use, but also a default discount rate (which then is just overridden)?

**Comment by MikePiper** [comment]
What would you think about the following?

When the calculator is loaded *without* a discount rate in the URL, there's
a message indicating that the default discount rate is 20-year TIPS yield.

When the calculator is loaded *with* a discount rate in the URL, there's a
message along the lines of "The calculator usually uses the 20-year TIPS
yield (which is currently ___%) as the default discount rate. But the page
was loaded with a URL that included a discount rate of  ___%."

And then the defaultDiscountRate variable is just a fixed number, which is
used when there is no internet. (Which I think is only for our own purposes
when running the calculator locally, without internet access?)

Would that satisfy the goals you are trying to satisfy?

-Mike

**Comment by brian-courts** [comment]
I'm honored that you seek my input. Your approach sounds quite reasonable. 

The user needs internet access to start the online calculator, so in almost all cases that access will provide the TIPS rate. The defaultDiscountRate COULD be the fixed number (currently 1) if there is an error getting the TIPS rate from www.quandl.com. It's good that you provided that backup.

Should we 
(a) keep the discount rate description as it is (dollar today, etc.), or
(b) make it very brief on the input page, with a link to a longer explanation on the About page?

if (b), the About page Discount Rate explanation could be combined with the Inflation topic. 

If I am reading the 2017 SSA Life Table correctly, it shows the male life expectancy at 62 is 82 (down from 83, as shown on the About page), and the female is down to 83 from 86. I could revise those, also.

Once you let me know, I'll get to work implementing the changes.     On Wednesday, July 15, 2020, 09:38:59 AM CDT, MikePiper <notifications@github.com> wrote:  
 
 

What would you think about the following?

When the calculator is loaded *without* a discount rate in the URL, there's
a message indicating that the default discount rate is 20-year TIPS yield.

When the calculator is loaded *with* a discount rate in the URL, there's a
message along the lines of "The calculator usually uses the 20-year TIPS
yield (which is currently ___%) as the default discount rate. But the page
was loaded with a URL that included a discount rate of ___%."

And then the defaultDiscountRate variable is just a fixed number, which is
used when there is no internet. (Which I think is only for our own purposes
when running the calculator locally, without internet access?)

Would that satisfy the goals you are trying to satisfy?

-Mike

**Comment by MikePiper** [comment]
Thanks for the catch regarding the change in median life expectancy from
the new SSA table.

I just pushed some code that provides the messages discussed above
regarding discount rate source. (Haven't yet deployed it to the live
version of the site.)

Basically I deleted the defaultDiscountRate variable and added a new one,
so now there are just three separate variables (tipsDiscountRate,
urlDiscountRate, and defaultDiscountRateIfError) that are used as necessary.

Does that work for you in terms of achieving the goals you had for that
section? If so I'll go ahead and deploy it.

Thanks again for your efforts. The interface used to get super overwhelming
as soon as "advanced" was checked -- this is much better.

-Mike

**Comment by brian-courts** [comment]
I've synced my 'master' repo to yours and ran it. Your adjustments to my proposals look very good. Just a couple of minor points:

I suggest that, in the 'additional inputs' list, you show 'Discount Rate other than the default (xxx%), where xxx is the initial discount rate (whether from URL or TIPS) . That would help the user decide whether to click the box or take the initial value.

When the discount rate comes from the URL, you show '(which is currently (TIPS)%)' right after the TIPS link. I think it would be useful to also show that info when the default is TIPS, and starts out in the input box. After a user enters other rates and scenarios to see effects, she might wonder, "Now, what did Mike say the TIPS rate is?" 

- Brian     On Wednesday, July 15, 2020, 04:22:17 PM CDT, MikePiper <notifications@github.com> wrote:  
 
 

Thanks for the catch regarding the change in median life expectancy from
the new SSA table.

I just pushed some code that provides the messages discussed above
regarding discount rate source. (Haven't yet deployed it to the live
version of the site.)

Basically I deleted the defaultDiscountRate variable and added a new one,
so now there are just three separate variables (tipsDiscountRate,
urlDiscountRate, and defaultDiscountRateIfError) that are used as necessary.

Does that work for you in terms of achieving the goals you had for that
section? If so I'll go ahead and deploy it.

Thanks again for your efforts. The interface used to get super overwhelming
as soon as "advanced" was checked -- this is much better.

-Mike

**Comment by MikePiper** [comment]
Both suggestions (displaying default discount rate at the top, and displaying TIPS yield in the explanation box when a URL discount rate is used) have been included in the latest push.

I also went ahead and deployed it. I'm confident that people will appreciate your improvement.

**Comment by brian-courts** [comment]
Thanks. 

I'd call it "our" improvement, considering your helpful adjustments.
     On Thursday, July 16, 2020, 09:08:50 AM CDT, MikePiper <notifications@github.com> wrote:  
 
 


Both suggestions (displaying default discount rate at the top, and displaying TIPS yield in the explanation box when a URL discount rate is used) have been included in the latest push.

I also went ahead and deployed it. I'm confident that people will appreciate your improvement.
